Enhance Module 6: Improve Task Instructions and Code Examples (#19)
udplabs/auth0-ai: public/assets/docs/demo-platform/06-module-06.md

2. ~~Guard against a missing Auth0AI client (*defensive coding*)~~ _<span style='color: green; font-variant: small-caps'>← Done for you</span>_

3. **Add the Required Imports**: At the top of the file, add:
   ```typescript
   import type { AuthorizerToolParameter, TokenSet } from '@auth0/ai';
   import {
   	AccessDeniedInterrupt,
   	UserDoesNotHavePushNotificationsInterrupt,
   } from '@auth0/ai/interrupts';

>    ```

Reviewer: ```suggestion
3. Add the custom scope (aka 'permission') we created earlier to the existing scopes array.

	> [!TIP]
	> **Scope Configuration Best Practices:**
	> 
	> - Always include baseline OIDC scopes: `openid`, `profile`, `email`
	> - Add custom API scopes like `create:transfer` for specific permissions
	> - In production, consider loading scopes from environment variables:
	>   ```typescript
	>   const scopes = process.env.AUTH0_API_SCOPES?.split(',') || ['openid', 'profile', 'email'];
	>   ```
	> - Scopes should match exactly what you configured in the Auth0 API

====== Move to 1. (line 371) and adjust numbering accordingly. =====
1. Ensure the necessary imports have been added to the file: 
   ```typescript
   import type { AuthorizerToolParameter, TokenSet } from '@auth0/ai';
   import {
   	AccessDeniedInterrupt,
   	CIBAInterrupt,
   	UserDoesNotHavePushNotificationsInterrupt,
   } from '@auth0/ai/interrupts';
   ```
```

4. **Update the Guard Condition**: Change the return statement in the guard to return the tool instead of nothing:
   ```typescript
   if (!auth0AI) {
   	console.warn('Auth0AI client not initialized!');

>    	return tool; // Instead of just 'return;'

Reviewer: @jansinger what would the benefit be of returning the tool in this situation? Are you suggesting we modify the behavior so that transferFunds() works correctly without any authorization?

Author: @eatplaysleep this avoids a type error in tool-registry.ts, see screenshot. You could change the ToolSet type to prevent this as well.
<img width="798" height="414" alt="Bildschirmfoto 2025-12-09 um 11 50 26" src="https://github.com/user-attachments/assets/a2f7f61d-1f35-4bb3-9e98-22ad3c5f3551" />

   }
   ```

5. **Add the Custom Scope**: Update the scopes array to include the transfer permission:
   ```typescript
   const scopes = ['openid', 'profile', 'email', 'create:transfer'];
   ```

> 	<br>

Reviewer: ```suggestion
5. Insert the <kbd>audience</kbd> value we created earlier.
```

	> [!TIP]
	> **Scope Configuration Best Practices:**
	> 
	> - Always include baseline OIDC scopes: `openid`, `profile`, `email`
	> - Add custom API scopes like `create:transfer` for specific permissions
	> - In production, consider loading scopes from environment variables:
	>   ```typescript
	>   const scopes = process.env.AUTH0_API_SCOPES?.split(',') || ['openid', 'profile', 'email'];
	>   ```

> 	> - Scopes should match exactly what you configured in the Auth0 API

Reviewer: Recommend moving to step 3. 
```suggestion
6. Enhance <kbd>onAuthorizationRequest</kbd> by plugging in our custom <kbd>handleOnAuthorize</kbd> helper function.
```

6. **Implement the Full Configuration**: Replace the commented-out implementation with the complete `withAsyncUserConfirmation` call:
   ```typescript
   return auth0AI.withAsyncUserConfirmation({
   	scopes,
   	userID: async () => {
   		const user = await getUser();
   		return user.sub;
   	},
   	audience: 'http://localhost:3000/api/accounts/transfers',
   	onAuthorizationRequest: handleOnAuthorize(writer),
   	onUnauthorized: (e) => {
   		// Handle different interrupt types...
   	},
   	...options,
   })(tool);

>    ```

Reviewer: I understand the suggestion but, in general, we have attempted to take a different approach with this lab away from simply copy/pasting code to instead users having to write their own code (to better learn). 

_This lab is **not** an 'intro' lab and is intended to be taken by experienced developers._

Aiya is designed to be able to provide the "final code" on request (has access to it) and also trained to offer progressively more detailed hints if the user is struggling. 

```suggestion

```

Author: Understood. It is a little bit confusing that the other steps are mainly copy and paste. Maybe add a hint that now a different apporach is used?

<br>

7. **Implement Error Handling**: In the `onUnauthorized` callback, handle different interrupt types:

   ```typescript
   onUnauthorized: (e) => {
   	if (e instanceof AccessDeniedInterrupt) {
   		return {
   			status: 'denied',
   			dataCount: 0,
   			message: '🚫 **Berechtigung verweigert**\n\nDie angeforderte Aktion wurde von Ihnen abgelehnt...',
   			error: e,
   			reason: 'user_denied_authorization',
   			canRetry: true,
   			nextSteps: [
   				'Stellen Sie sicher, dass Sie die Aktion wirklich durchführen möchten',
   				'Versuchen Sie es erneut mit dem gleichen Befehl',
   				'Kontaktieren Sie den Support, falls Sie Probleme haben',
   			],
   		};
   	}

   	if (e instanceof UserDoesNotHavePushNotificationsInterrupt) {
   		return {
   			status: 'error',
   			dataCount: 0,
   			message: 'MFA Push-Benachrichtigungen sind nicht eingerichtet. Ich starte das Enrollment für Sie...',
   			error: e,
   			requiresMFAEnrollment: true,
   			forceAction: 'enroll-mfa-push',
   			autoExecute: true,
   		};
   	}

   	// Default error handling
   	return {
   		status: 'error',
   		dataCount: 0,
   		message: e.message,
   		error: e,
   	};
   }

>    ```

Reviewer: Definitely open to improved error handling and internationalization but, similar to previous comment, we have purposefully avoided a copy/paste approach with the lab. 

```suggestion

```

	<br>

	> [!TIP]
	> **Advanced Error Handling Options:**
	>
	> - **Custom Error Messages**: You can customize messages based on the specific error or user context
	> - **Retry Logic**: Add `canRetry: true` with custom retry mechanisms
	> - **Conditional Actions**: Use different `forceAction` values based on error type
	> - **User Guidance**: Provide `nextSteps` arrays with actionable instructions

> 	> - **Streaming Updates**: Use the `writer` to stream real-time error explanations to the UI

Reviewer: I like the suggestion for more advanced error handling, but it could end up breaking things or changing the flow considerably which could make the lab more difficult to troubleshoot. 

Although the agent/app _technically_ should be able to handle these concepts, the outcome is not guaranteed and it may interfere with other components of the app. 
 
```suggestion
    > *Not sure what to do here?*
	>
	> This wrapper returns to a *tool*, which then returns to the *streaming function*.
	>
	> How does the *tool* handle/return errors? 🤔
```

	> - **Retry Logic**: Add `canRetry: true` with custom retry mechanisms
	> - **Conditional Actions**: Use different `forceAction` values based on error type
	> - **User Guidance**: Provide `nextSteps` arrays with actionable instructions
	> - **Streaming Updates**: Use the `writer` to stream real-time error explanations to the UI
	>
	> Refer to **Step 4**: try passing an *async function* that returns an error object.
	> 	- The goal is to ensure that an error is returned *gracefully* if it occurs and does not 'throw' or *halt* the current action entirely.
	> 	- If the function halts processing the Agent will not be able to properly triage and find alternatives solutions.

	<br>

8. Now, spread the incoming options so they are passed along to <kbd>withAsyncUserConfirmation</kbd>.

	Not sure what a "spread" is? *Ask Aiya*!

	```diff
	- // 	...options, /** 👀 ✅ Step 8: The Auth0AI wrapper spreads the same options as our wrapper! TypeScript interface to the rescue? 🧐 */
	+ ...options,

9.  Lastly, make sure the tool being wrapped is *actually injected*!
	```diff
	- // })(tool) /** ✅ Step 9: Don't forget to inject the `tool` being wrapped! */;
	+ })(tool);

> 	```

Reviewer: Keep

> 	> ```

Reviewer: This structure does not actually match the types defined in the app so it would likely cause some major issues without making code changes elsewhere. 

```suggestion
	> Refer to **Step 4**: try passing an *async function* that returns an error object.
	> 	- The goal is to ensure that an error is returned *gracefully* if it occurs and does not 'throw' or *halt* the current action entirely.
	> 	- If the function halts processing the Agent will not be able to properly triage and find alternatives solutions.

	<br>

8. Now, spread the incoming options so they are passed along to <kbd>withAsyncUserConfirmation</kbd>.

	Not sure what a "spread" is? *Ask Aiya*!

	```diff
	- // 	...options, /** 👀 ✅ Step 8: The Auth0AI wrapper spreads the same options as our wrapper! TypeScript interface to the rescue? 🧐 */
	+ ...options,
```

<span style="font-variant: small-caps; font-weight: 700">Setup</span>

> - From your code editor, open `lib/ai/tools/transfer-funds.ts`.

Reviewer: 👍🏻 this has already been updated in the demo platform version. This version is slightly outdated.

<span style="font-variant: small-caps; font-weight: 700">Steps</span>

1. **Add Required Imports**: At the top of the file, add the necessary imports:
   ```typescript
   import { withAsyncAuthorization } from '@/lib/auth0/ai/with-async-authorization';
   import { getCIBACredentials } from '@auth0/ai-vercel';

>    ```

Reviewer: ```suggestion
  1. Ensure the necessary imports have been added to the file:
     ```typescript
     import { withAsyncAuthorization } from '@/lib/auth0/ai/with-async-authorization';
     import { getCIBACredentials } from '@auth0/ai-vercel';
     import { tool, type UIMessageStreamWriter } from 'ai';
     ```
```

2. **Transform the Tool Export**: Currently, `transferFunds` is exported as a simple tool. You need to wrap it with `withAsyncAuthorization`. Replace the entire export from:
   ```typescript
   export const transferFunds = tool<...>({
   ```
   to:
   ```typescript
   export const transferFunds = (writer?: UIMessageStreamWriter) =>
   	withAsyncAuthorization({
   		tool: tool<...>({

>    ```

Reviewer: Avoiding copy/paste approach. 

```suggestion
2. Wrap <kbd>transferFunds</kbd> with <kbd>withAsyncAuthorization</kbd>.
	You will need to:
      - import the function from the `lib/auth0/ai` directory.
      - *wrap the tool* -- instead of simply returning it, pass it as the <kbd>tool</kbd> parameter of <kbd>withAsyncAuthorization</kbd>.
      - <kbd>transferFunds</kbd> should ultimately *still return* the original tool.
      - <kbd>bindingMessage</kbd> can be a simple string like `Please approve the transfer.`
```

	<br>

	> [!TIP]
	> **Higher-Order Function Pattern:**
	> 
	> This creates a "factory function" that:
	> - Accepts an optional `writer` parameter for streaming updates
	> - Returns a wrapped tool with authorization capabilities
	> - Maintains the original tool's TypeScript interface
	> 
	> **Why this pattern?** It allows the same tool to be used with or without streaming capabilities while maintaining type safety.

> 

Reviewer: 👍🏻 I am in favor. There was a similar tip in the original version but it ended up getting removed.

Reviewer: I am not sure how we missed the fact that 7 and 8 are redundant. I think originally they were combined. I'm going to refactor them entirely and incorporate some of your wording suggestions.

3. **Add Authorization Parameters**: At the end of the `withAsyncAuthorization` call, add:
   ```typescript
   		}),
   		writer,
   		bindingMessage: 'Please approve the transfer',
   	});

>    ```

Reviewer: ```suggestion
3. Be sure to pass <kbd>bindingMessage</kbd> to <kbd>withAsyncAuthorization</kbd>: 
```

	> [!NOTE]
	>
	> ***What is <kbd>bindingMessage</kbd>?***
	>
	> When using the Auth0 Guardian SDK this message can be displayed to the user in order to provide context about the request. It is *not* used in our demo but still required.

4. **Retrieve CIBA Credentials**: In the tool's `execute` function, replace the empty `Authorization` header with:
   ```typescript
   const token = getCIBACredentials();
   // ...
   Authorization: `Bearer ${token?.accessToken}`,

>    ```

Reviewer: ```suggestion
4. In **step one** you imported <kbd>getCIBACredentials</kbd>. Now use it to retrieve an <kbd>accessToken</kbd>. Ensure the token is sent in the tool's <kbd>fetch</kbd> call as the <kbd>Authorization</kbd> header
```

	<br>

	> [!TIP]
	> **Understanding CIBA Credentials:**
	> 
	> - `getCIBACredentials()` returns the fresh access token obtained through the CIBA flow
	> - This token contains the `create:transfer` scope we configured earlier
	> - The token is ephemeral and specific to this authorization request
	> - Always use optional chaining (`?.`) when accessing token properties for safety

5. **Update the Tool Description**: Change the description to remove the manual confirmation requirement:

>    Change the text from:

Reviewer: ```suggestion
   Change the following text to explicitly **not** require confirmation. There is no "right" way to word this -- you decide. 
```

   ```
   Always confirm the details of the transfer with the user before continuing.
   ```
   to:
   ```
   DO NOT require confirmation from the user -- they will confirm via push notification.

>    ```

Reviewer: ```suggestion

```

	> [!TIP]

> 	> When instructing <abbr title='large language models'>LLMs</abbr> be explicit but concise. The push notification will handle user confirmation automatically.

Reviewer: ```suggestion
	> Instructing an <abbr title='large language models'>LLM</abbr> is like training someone to perform the task. Sometimes adding an explanation as to *why* the model should do/not do something can be helpful (i.e. *the user will be getting a push notification instead*).
	>
	> When instructing <abbr title='large language models'>LLMs</abbr> be explicit but concise. 
```

	***How?***

	This was already accomplished in Task 7! By transforming the export to:
	```typescript
	export const transferFunds = (writer?: UIMessageStreamWriter) => withAsyncAuthorization({...

> 	```

Reviewer: Will refactor all of this.

<span style="font-variant: small-caps; font-weight: 700">Steps</span>

1. Open `lib/ai/tool-registry.ts`.
2. Find the line with `transferFunds` and update it to call the function:

> 	```diff

Reviewer: Demo platform MD rendering engine sadly does not support `diff`. The official version was already updated.
```suggestion
	```ts
```

	- transferFunds /* ⚠️ TASK 9: Modify to call higher-order function (see `transfer-funds.ts`) */,
	+ transferFunds: transferFunds(), // call the higher-order function to obtain the tool
	```

	<br>

	> [!TIP]
	> **Tool Registry Patterns:**
	> 
	> - **Simple tools**: `toolName` (direct reference)
	> - **Factory functions**: `toolName()` (function call)
	> - **Parameterized tools**: `toolName(config)` (with configuration)
	> 
	> The registry calls `transferFunds()` without arguments, using the optional writer parameter pattern.

***But wait, there's more!***

3. Open `app/(chat)/api/chat/[id]/_handlers/post.ts`
4. Find the commented line in the initial tools setup and **remove** the comment:

> 	```diff

Reviewer: ```suggestion
	```ts
```

Nifty trick, *right*? 🤓

5. Now find the section where tools are re-initialized with the dataStream writer (around line 110). **Uncomment** the line with `dataStream` and **comment out** the line above it:

> 	```diff

Reviewer: ```suggestion
	```ts
```
